File: hooks/safety/ci-batch-check.py
#!/usr/bin/env python3
"""
CI Batch Check Hook - Warns about multiple pushes in short timeframe.

FAANG best practice: 1 logical unit = 1 push
Prevents CI queue explosion from incremental fixes.
"""

import json
import sys
import time
from pathlib import Path

# Config
PUSH_COOLDOWN_MINUTES = 10
PUSH_HISTORY_FILE = Path("/tmp/claude_push_history.json")
MAX_RAPID_PUSHES = 2  # Warn after this many pushes within cooldown


def load_push_history() -> list:
    """Load push timestamps from history file."""
    if PUSH_HISTORY_FILE.exists():
        try:
            with open(PUSH_HISTORY_FILE) as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            return []
    return []


def save_push_history(history: list) -> None:
    """Save push timestamps to history file."""
    # Keep only last hour of history
    cutoff = time.time() - 3600
    history = [ts for ts in history if ts > cutoff]
    with open(PUSH_HISTORY_FILE, "w") as f:
        json.dump(history, f)


def get_recent_pushes(history: list, minutes: int) -> int:
    """Count pushes within the last N minutes."""
    cutoff = time.time() - (minutes * 60)
    return sum(1 for ts in history if ts > cutoff)


def main():
    # Read hook input
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        sys.exit(0)

    tool_input = input_data.get("tool_input", {})
    command = tool_input.get("command", "")

    # Only check git push commands
    if "git push" not in command and "git push" not in command.replace("&&", " "):
        sys.exit(0)

    # Load history
    history = load_push_history()
    recent_count = get_recent_pushes(history, PUSH_COOLDOWN_MINUTES)

    # Record this push attempt
    history.append(time.time())
    save_push_history(history)

    # Check if too many rapid pushes
    if recent_count >= MAX_RAPID_PUSHES:
        # Output warning but don't block
        output = {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "suppressToolUse": False,
                "message": f"[CI Batch Warning] {recent_count + 1} pushes in {PUSH_COOLDOWN_MINUTES}min. "
                f"Consider batching fixes into 1 logical unit per push (FAANG best practice).",
            }
        }
        print(json.dumps(output))
        sys.exit(0)

    sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/safety/ci-batch-check.js
#!/usr/bin/env node
/**
 * CI Batch Check Hook - Warns about multiple pushes in short timeframe
 *
 * FAANG best practice: 1 logical unit = 1 push
 * Prevents CI queue explosion from incremental fixes.
 *
 * Configurable via ~/.claude/ci-config.json
 *
 * Returns: { decision: "warn", message: "..." } or {}
 */

const path = require('path');
const fs = require('fs');
const { readStdinJson, output, getHomeDir } = require(path.join(__dirname, '..', '..', 'lib', 'utils.js'));

// Default configuration
const DEFAULT_CONFIG = {
  enabled: true,
  pushCooldownMinutes: 10,
  maxRapidPushes: 2,
  trackForcePushes: true,
  maxForcePushesPerHour: 1
};

// History file location
const PUSH_HISTORY_FILE = path.join(getHomeDir(), '.claude', 'metrics', 'push_history.json');

/**
 * Load configuration from ~/.claude/ci-config.json
 */
function loadConfig() {
  const configPath = path.join(getHomeDir(), '.claude', 'ci-config.json');
  try {
    if (fs.existsSync(configPath)) {
      const userConfig = JSON.parse(fs.readFileSync(configPath, 'utf8'));
      return { ...DEFAULT_CONFIG, ...userConfig };
    }
  } catch (err) {
    console.error(`Warning: Could not load CI config: ${err.message}`);
  }
  return DEFAULT_CONFIG;
}

/**
 * Load push history from temp file
 */
function loadPushHistory() {
  try {
    if (fs.existsSync(PUSH_HISTORY_FILE)) {
      const data = JSON.parse(fs.readFileSync(PUSH_HISTORY_FILE, 'utf8'));
      return {
        pushes: data.pushes || [],
        forcePushes: data.forcePushes || []
      };
    }
  } catch (err) {
    // Ignore errors, return empty history
  }
  return { pushes: [], forcePushes: [] };
}

/**
 * Save push history to temp file
 */
function savePushHistory(history) {
  try {
    const now = Date.now();
    const oneHourAgo = now - (60 * 60 * 1000);

    // Keep only last hour of history
    history.pushes = history.pushes.filter(ts => ts > oneHourAgo);
    history.forcePushes = history.forcePushes.filter(ts => ts > oneHourAgo);

    // Ensure directory exists before writing
    const dir = path.dirname(PUSH_HISTORY_FILE);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    fs.writeFileSync(PUSH_HISTORY_FILE, JSON.stringify(history), 'utf8');
  } catch (err) {
    console.error(`Warning: Could not save push history: ${err.message}`);
  }
}

/**
 * Count events within the last N minutes
 */
function countRecentEvents(timestamps, minutes) {
  const cutoff = Date.now() - (minutes * 60 * 1000);
  return timestamps.filter(ts => ts > cutoff).length;
}

/**
 * Check if command is a git push
 */
function isGitPush(command) {
  return /git\s+push/.test(command);
}

/**
 * Check if command is a force push
 */
function isForcePush(command) {
  return /git\s+push\s+.*\s(-f|--force)(\s|$)/.test(command);
}

/**
 * Get branch being pushed to
 */
function getPushBranch(command) {
  // Try to extract branch from command
  const match = command.match(/git\s+push\s+\S+\s+(\S+)/);
  if (match) {
    return match[1];
  }
  // Default to current branch indicator
  return 'current branch';
}

async function main() {
  try {
    const config = loadConfig();

    if (!config.enabled) {
      output({});
      process.exit(0);
    }

    const input = await readStdinJson();

    const toolName = input.tool_name || '';
    const toolInput = input.tool_input || {};
    const command = toolInput.command || '';

    // Only check Bash commands with git push
    if (toolName !== 'Bash' || !isGitPush(command)) {
      output({});
      process.exit(0);
    }

    const history = loadPushHistory();
    const now = Date.now();
    const isForce = isForcePush(command);
    const branch = getPushBranch(command);

    // Check force push limits
    if (isForce && config.trackForcePushes) {
      const forcePushCount = countRecentEvents(history.forcePushes, 60); // Last hour

      if (forcePushCount >= config.maxForcePushesPerHour) {
        output({
          decision: 'warn',
          message: `[CI Batch Warning] ${forcePushCount + 1} force pushes in the last hour.\n` +
            `Force pushing repeatedly can disrupt CI pipelines and collaborators.\n` +
            `Consider squashing your fixes into a single force push.`
        });

        // Record this force push
        history.forcePushes.push(now);
        savePushHistory(history);
        process.exit(0);
      }

      // Record force push
      history.forcePushes.push(now);
    }

    // Check rapid push limits
    const recentPushCount = countRecentEvents(history.pushes, config.pushCooldownMinutes);

    // Record this push
    history.pushes.push(now);
    savePushHistory(history);

    if (recentPushCount >= config.maxRapidPushes) {
      output({
        decision: 'warn',
        message: `[CI Batch Warning] ${recentPushCount + 1} pushes in ${config.pushCooldownMinutes} minutes.\n` +
          `Branch: ${branch}\n\n` +
          `Consider batching fixes into 1 logical unit per push (FAANG best practice).\n` +
          `This helps:\n` +
          `  - Reduce CI queue congestion\n` +
          `  - Create cleaner git history\n` +
          `  - Make code reviews easier`
      });
      process.exit(0);
    }

    // All good
    output({});
    process.exit(0);

  } catch (err) {
    // On error, fail open (allow operation)
    console.error(`CI batch check error: ${err.message}`);
    output({});
    process.exit(0);
  }
}

main();
